### PR TITLE
Configurable deployment modes

### LINKED ISSUE #11: Add a 'copy' deployment mode for cargo-screeps
I often find myself wanting to deploy to a local sandbox server that I'm running, and sometimes can't be bothered to set up the authentication mod that allows for the http upload functionality to work.

It would be nice to have an option to instead have the build artifacts copied to the directory that the screeps client watches and uploads for you.

The config could be something like:

```toml
mode = "copy" # Default: copy, valid: copy|upload
branch = "default"

[upload]
username = "user"
password = "pass"
# optional:
# ptr = false
# hostname = "localhost"
# ssl = true
# port = 443

[copy]
dest = "/path/to/screeps/client/scripts" # path to directory containing the branch 
                                         # directories, configured branch should be
                                         # appended
prune = true # whether or not to clear the destination of extra files
```

Edit: This is probably something that I'd be willing to work on myself.

### PR DESCRIPTION
**Pull request by daboross**
This is a PR with my changes on top of @jechase's from #12 

Closes #12. Fixes #11.

### CODE REVIEW

File: cargo-screeps/src/copy.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    let target_dir = root.join("target");
+
+    let mut deployed: HashSet<PathBuf> = HashSet::new();
+
+    for entry in fs::read_dir(target_dir)? {
+        let entry = entry?;
+        let path = entry.path();
+
+        if let (Some(name), Some(extension)) = (path.file_name(), path.extension()) {
+            if extension == "js" {
```
Comment (ASalvail) [inline cargo-screeps/src/copy.rs]: If at some point the user changes the configured file name, they'd end up with two "main" files and wasm files copied over.

Shouldn't that refer directly to the build config specific file names?

File: cargo-screeps/src/config.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    pub branch: String,
+    pub ssl: bool,
+    pub port: i32,
+    pub ptr: bool,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct CopyConfiguration {
+    pub destination: PathBuf,
+    pub branch: String,
```
Comment (ASalvail) [inline cargo-screeps/src/config.rs]: Unless I'm mistaken, the `branch` isn't used (at least, not in `copy.rs`)

Comment (daboross) [reply cargo-screeps/src/config.rs]: I think it should be, at least. It's used as a sub-directory under destination to stick files in.

File: cargo-screeps/src/copy.rs (hunk 1 above)
Comment (daboross) [reply cargo-screeps/src/copy.rs]: If it should, it should change in both copy and upload. This preserves the current behavior of uploading files.

I.. think the purpose is to allow people to include extra javascript files if they want to?

Build will fail anyways if the project name is changed and `cargo clean` isn't run.

File: cargo-screeps/src/config.rs (hunk 2 above)
Comment (ASalvail) [reply cargo-screeps/src/config.rs]: Indeed, I had missed that. Nevermind!

**Comment by daboross** [comment]
I think this should be fairly complete now!

There isn't backwards compatibility with the old configuration, but it shouldn't be hard to migrate - it's just adding an `[upload]` line. Instead `cargo screeps` will now warn on unused configuration values, and `cargo screeps upload` will complain if there isn't an `[upload]` section.

I'll update `screeps-starter-rust` after this is merged and released.

**Comment by daboross** [comment]
Cool!

I've been trying it out locally too and it seems to be pretty reliable now. I'm not writing integration tests for this tool but I think that's probably alright, those can come later.

**Pull request merged**